### ISSUE TITLE
Network Slimming Re-implementation For Sparse Training

### ISSUE THREAD

**Issue by rogercmq** [rogercmq/NetworkCompressionPaperNotes#2]
# Network Slimming 存在的问题:

1. 这种加速 inference 的方法需要fine-tune, 但是在我们的场景下sparse network是不断稀疏的所以没有微调这种说法。

2. 如果直接拿来做 training phase 的网络压缩, 我目前跑的实验是每20个回合减一次, 然后用剪枝后的比较稀疏的模型继续训练, 反复迭代整个过程直到模型收敛。

3. 但是这种方法（pruning for inference）是有问题的：

    - 具体在哪个 epoch 去剪枝是empirical and case-by-case 的，也就是说我现在跑的这个network slimming 并不符合我们动态剪枝的要求，我必须在代码里指明哪个epoch作为breakpoint去做剪枝。
    - 如果确定在某 epoch 剪枝，全局剪枝率(在 network slimming 中是)作为超参，如果剪枝率太大则 accuracy drop 惨不忍睹，剪枝率太小则收敛时间太长。稀疏化训练的 robustness 就会有问题。例如：

        > i. After training 20 epochs, VGG11 model accuracy on CIFAR10: 87.6%
        > ii. Pruning... 
        > iii. After pruning (剪枝率0.2), VGG11 model accuracy on CIFAR10: **70.4%**
        > ...
    - 压缩后的模型不能继续用之前 optimizer 对应的值 (weight decay, learning rate, etc.) 更新参数，因为网络架构改变了，原来炼丹的超参不一定是最好的选择。

# 目前的工作:

0. 写一个 pytorch 函数，把原模型压缩到一个新模型里面。可参考 Xiaoxing Wang 的代码。

1. 复现 Learning Structured Sparsity in Deep Neural Networks 

    > NIPS 2016, Yiran Chen Group
    > 最早提出在训练过程中根据 group LASSO 结构化稀疏的一篇文章
    > 源码是 Caffe 但是
    > Github Link: [Caffe Source Code](https://github.com/wenwei202/caffe/tree/scnn)
    > **正在做的工作**: re-implement by pyTorch

**Comment by rogercmq** [comment]
# Rethinking Structural-Sparsity Network Training
### I. 训练方法的炼丹
- 前10轮不剪枝
- 最后5轮不剪枝(considered as finetune)
- 中间层每隔5轮剪枝一次

对应的[10,5,5]可能得重新考虑一下，现在的收敛速度还是慢，而且accuracy不能恢复到pretrained acc

### II. Optimizer的问题
Adam (not SGD!)

### III.本周必须做的实验
- prune by mask (变conv不变bn)
- prune by double masking (变conv不变bn, 直接把剪枝之后的东西 deploy 到新模型上)

考虑 ThiNet 的剪枝思路，对于 ResNet 的block保留最后一个layer的BN，不然skip-shortcut会匹配不上。
